pennylane/ftqc/qubit_graph.py

```diff
@@ -0,0 +1,601 @@
+# Copyright 2025 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This module contains the data structures to represent qubits using a hierarchical memory model
+and defines the API to access and manipulate these data structures.
+"""
+
+import warnings
+from typing import Any, Optional, Union
+
+import networkx as nx
+
+# Fail-safe for algorithms that traverse the nested graph structure
+# In other words, we should never expect more than this many layers of nested QubitGraphs
+MAX_TRAVERSAL_DEPTH = 100
+
+
+class QubitGraph:
+    """A class to represent a hierarchical qubit memory model as nested graphs of qubits.
+
+    A QubitGraph is a qubit that contains a graph of underlying qubits, where each underlying qubit
+    is itself a QubitGraph. This representation allows for nesting of lower-level qubits with
+    arbitrary depth to allow easy insertion of arbitrarily many levels of abstractions between
+    logical qubits and physical qubits.
+
+    Args:
+        id (Any): An identifier for this QubitGraph object.
+        graph (graph-like, optional): The graph to use as the QubitGraph's underlying qubits.
```

Comment:
Initially when trying to interact with this, I made a nested `networkx` graph and passed it in, which sort of looks like it works initially, but actually uses the sub-graphs and labels and leaves the nodes that "should" contain them uninitialized.

I don't think we should iterate over this whole thing and add a check, because that sounds expensive, but I do think we should note in the docstring that this input graph can't be nested, and include an example in the docstring of the correct procedure for creating a nested graph (by repeatedly setting nodes to be single-layer graphs at various levels).